torch/_subclasses/functional_tensor.py

```diff
@@ -0,0 +1,259 @@
+            any_functional_inputs = True
+            return x.elem
+
+        args_unwrapped, kwargs_unwrapped = pytree.tree_map_only(
+            FunctionalTensor, unwrap, (args, kwargs)
+        )
+
+        if not self.is_active:
+            assert not any_functional_inputs
+            return func(*args_unwrapped, **kwargs)
+
+        # Expectation: functionalization should not **already** be enabled above our mode.
+        # Why would that be bad? when we return a FunctionalTensor here, we don't want functionalization
+        # to run above this mode and further wrap that output in **another** C++ FunctionalTensorWrapper.
+        is_included = torch._C._dispatch_tls_is_dispatch_key_included(
+            torch._C.DispatchKey.Functionalize
+        )
+        is_excluded = torch._C._dispatch_tls_is_dispatch_key_excluded(
+            torch._C.DispatchKey.Functionalize
+        )
+        assert is_excluded or not is_included
+        # All we want to do here is re-use the existing C++ functionalization logic.
+        # This requires swizzling our TLS dispatch keys so that the Functionalize key is active.
+        with torch._C._SetExcludeDispatchKeyGuard(
+            torch._C.DispatchKey.Functionalize, False
+        ), torch._C._IncludeDispatchKeyGuard(torch._C.DispatchKey.Functionalize):
+            try:
+                # By default for python functionalization (for AOTAutograd), we reapply views.
+                torch._functionalize_enable_reapply_views(True)
+                outs_unwrapped = func(*args_unwrapped, **kwargs_unwrapped)
+                pytree.tree_map_only(torch.Tensor, assert_is_functional, outs_unwrapped)
+
+                outs_wrapped = pytree.tree_map_only(torch.Tensor, wrap, outs_unwrapped)
+            finally:
+                torch._disable_functionalization()
+
+        is_included = torch._C._dispatch_tls_is_dispatch_key_included(
+            torch._C.DispatchKey.Functionalize
+        )
+        is_excluded = torch._C._dispatch_tls_is_dispatch_key_excluded(
+            torch._C.DispatchKey.Functionalize
+        )
+        assert is_excluded or not is_included
+
+        # Wrapper tensor subclasses do not have correct aliasing info! Use this util to manually correct the output aliasing.
+        # inplace ops like `aten.add_()` are expected to return inputs **directly**, instead of creating fresh tensor objects.
+        # Use this util to figure out the right thing to return.
+        # If none of our inputs were wrapped, then we have no FunctionalTensor outputs that we need to fix up storages for.
+        return return_and_correct_aliasing(func, args, kwargs, outs_wrapped)
+
+
+@contextlib.contextmanager
+def maybe_disable_functional_mode():
+    maybe_func_mode = torch._C._unset_dispatch_mode(
+        torch._C._TorchDispatchModeKey.FUNCTIONAL
+    )
+    try:
+        yield
+    finally:
+        if maybe_func_mode is not None:
+            torch._C._set_dispatch_mode(maybe_func_mode)
```

Comment:
In the other PR, you don't use this context manager, instead you enable/disable the mode without removing it from the stack. Is there are a particular reason to do it that way?

Reply:
nope - in retrospect, an explicit enable/disable (by adding an extra `is_active` field on the mode) feels wrong. We already have an API now for setting/unsetting the mode, so using that here feels like the right thing.